dag_in_context/src/optimizations/loop_invariant.egg

```diff
@@ -1,93 +1,84 @@
 ;; Loop Invariant
 
 ;; bool: whether the term in the Expr is an invariant.
-(function is-inv-Expr (Expr Expr) bool :merge (or old new))
-(function is-inv-ListExpr (Expr ListExpr) bool :merge (or old new))
-
-;; in default, when there is a find, set is-inv to false
-(rule ((BodyContainsExpr loop term)
-       (= loop (DoWhile inputs pred_out)))
-      ((set (is-inv-Expr loop term) false)) :ruleset always-run)
-(rule ((BodyContainsListExpr loop term)
-       (= loop (DoWhile inputs pred_out)))
-      ((set (is-inv-ListExpr loop term) false)) :ruleset always-run)
+(relation is-inv-Expr (Expr Expr))
+(relation is-inv-ListExpr (Expr ListExpr))
 
 (relation is-inv-ListExpr-helper (Expr ListExpr i64))
-(rule ((BodyContainsListExpr loop list) 
-       (= loop (DoWhile inputs pred_out))) 
-      ((is-inv-ListExpr-helper loop list 0)) :ruleset always-run)
+(rule ((BodyContainsListExpr body list) ) 
+      ((is-inv-ListExpr-helper body list 0)) :ruleset always-run)
 
-(rule ((is-inv-ListExpr-helper loop list i)
-       (= true (is-inv-Expr loop expr))
-       (= expr (ListExpr-ith list i)))
-    ((is-inv-ListExpr-helper loop list (+ i 1))) :ruleset always-run)
+(rule ((is-inv-ListExpr-helper body list i)
+       (is-inv-Expr body (ListExpr-ith list i)))
+    ((is-inv-ListExpr-helper body list (+ i 1))) :ruleset always-run)
 
-(rule ((is-inv-ListExpr-helper loop list i)
+(rule ((is-inv-ListExpr-helper body list i)
        (= i (ListExpr-length list)))
-    ((set (is-inv-ListExpr loop list) true)) :ruleset always-run)
+    ((is-inv-ListExpr body list)) :ruleset always-run)
 
 
 (ruleset boundary-analysis)
-;; An Expr is on boundary when it is invariant and its parent is not
-;                       loop invariant-expr
-(relation boundary-Expr (Expr Expr))
-
-;; boundary for ListExpr's children
-(rule ((= true (is-inv-Expr loop expr))
-       (= false (is-inv-ListExpr loop list))
-       (= expr (ListExpr-ith list i)))
-      ((boundary-Expr loop expr)) :ruleset boundary-analysis)
-
-;; if a output branch/pred is invariant, it's also boundary-Expr
-(rule ((= true (is-inv-Expr loop expr))
-       (= loop (DoWhile in pred_out))
-       (= expr (Get pred_out i))) 
-      ((boundary-Expr loop expr)) :ruleset boundary-analysis)
-
-
-(function hoisted-loop (Expr Expr) bool :merge (or old new) )
-(rule ((= loop (DoWhile in pred_out)))
-      ((set (hoisted-loop in pred_out) false)) :ruleset always-run)
-
-(constructor InExtendedLoop (Expr Expr Expr) Assumption)
+(ruleset boundary-analysis-prep)
+
+;; We use a function so that we choose just one
+;; thing to hoist at a time
+;; This is an evil hack!
+;                   pred body invariant-expr
+(function to-hoist (Expr Expr) Expr :merge new)
```

Comment:
For a given loop, chooses a particular single invariant expression to hoist. This is a big performance improvement.